### PR TITLE
Refactor AST

### LINKED ISSUE #17: Refactor AST
I propose the following changes:

- [ ] Refactor `BodyAtom` to take an array of BodyTerms.
- [ ] Refactor the function in `GuardAtom` to take an array of values and to have an array of BodyTerm.
- [ ] Refactor `Constraint` to have an array of BodyPredicates.
- [ ] Refactor `HeadAtom`  to have an array of HeadTerms.
- [ ] Refactor `App` to have an array of HeadTerms.
- [ ] Refactor `SourceLocation` to use a 5-tuple and not a record.

Mostly to ease the translation from Flix to the AST.

As for the effects, I suggest to insert casts to pure whenever the computation is actually pure. In the future there might be an immutable array and then these casts could be removed.

### PR DESCRIPTION
**Pull request by bdahse**
Fixes #17 .

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Let me know when it is ready for review. I notice you have a lot of commented code.

**Comment by bdahse** [comment]
> 
> 
> Let me know when it is ready for review. I notice you have a lot of commented code.

You can review when you have time. The commented code was unused and related to semantic checks that we don't do (yet).

**Comment by magnus-madsen** [comment]
> > Let me know when it is ready for review. I notice you have a lot of commented code.
> 
> You can review when you have time. The commented code was unused and related to semantic checks that we don't do (yet).

Update or delete. I don't want commented code :)

**Comment by magnus-madsen** [comment]
Did you try it with the new Flix compiler branch?

File: main/Fixpoint/Ast/Constraint.flix [hunk 1]
```diff
@@ -16,15 +16,15 @@
...
-        pub def toString(c: Constraint[v]): String = match c {
-            case Constraint(head, Nil, _) => ToString.toString(head)
-            case Constraint(head, body, _) =>
-                let atoms = List.map(ToString.toString, body) |> String.intercalate(", ");
-                "${head} :- ${atoms}"
-        }
+        pub def toString(c: Constraint[v]): String =
+            use Fixpoint.commaSeparate;
+            match c {
+                case Constraint(head, [], _) => "${head}"
```
Comment (magnus-madsen) [inline main/Fixpoint/Ast/Constraint.flix:26]: Does this actually work? What notion of equality used for arrays? I think you should test the length instead.

Also, a fact/rule should end with a period :)

File: main/Fixpoint/Ast/ConstraintSystem.flix [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Fixpoint/Ast {
+    pub enum ConstraintSystem[v] {
```
Comment (magnus-madsen) [inline main/Fixpoint/Ast/ConstraintSystem.flix:18]: Let us call it `Datalog`. I like that name better.

File: main/Fixpoint/Solver.flix [hunk 3]
```diff
@@ -14,16 +14,6 @@
...
-        }, cs)
-
-    pub def nonPositivelyBoundVar(cs: List[Constraint[v]]): Bool =
-        List.forall(c -> match c {
-            case Constraint(_, body, _) => match polarityVars(body) {
-                case (positive, negative)  => negative `Set.isSubsetOf` positive
-            }
-        }, cs)
+namespace Fixpoint {
+    // TODO: implement solve, union, project, fold, entails.
```
Comment (magnus-madsen) [inline main/Fixpoint/Solver.flix:18]: OK, I guess this will come in next PR.

Comment (magnus-madsen) [reply main/Fixpoint/Solver.flix:18]: I suggest two solves: `solve(...)` and `solveWithStratification(...)`

**Review by magnus-madsen — COMMENTED** [review]
I added some minor comments - should be quick to fix and then we can merge.

File: main/Fixpoint/Solver.flix (hunk 3 above)
Comment (bdahse) [reply main/Fixpoint/Solver.flix:18]: > 
> 
> Did you try it with the new Flix compiler branch?

Not yet, but I will after I fix #15 . Then I should be able to rewrite my tests in "Flix Datalog".

**Comment by bdahse** [comment]
> 
> 
> Did you try it with the new Flix compiler branch?

Not yet, but I will after I fix #15 . Then I should be able to rewrite my tests in "Flix Datalog".

File: main/Fixpoint/Ast/ConstraintSystem.flix (hunk 2 above)
Comment (bdahse) [reply main/Fixpoint/Ast/ConstraintSystem.flix:18]: Alright, I will fix it in the next PR, this wasn't really supposed to be included in this one.

**Comment by magnus-madsen** [comment]
Excellent!

**Comment by magnus-madsen** [comment]
In other good news; I think I have figured out how to do boxing in a good way!

**Pull request merged**